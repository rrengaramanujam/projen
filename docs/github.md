# GitHub

By default, many projects are initialized with `GitHub` component to enabled GitHub as the default provider for CI/CD workflows. See <https://docs.github.com/en/actions> for more information.

The use of GitHub (and generating corresponding files in `.github`) can be disabled by specifying `github: false` in your project options.

## GitHub API access

Several workflows generated by projen use APIs that are not available through
the permissions of [GITHUB_TOKEN]. To use these workflows, you must provide
either a Personal Access Token (PAT) or a GitHub App to provide API access.

[GITHUB_TOKEN]: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

### Personal Access Token (classic)

Follow the [GitHub docs
instructions](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-personal-access-token-classic)
for creating a personal access token (classic).
When creating the classic PAT, grant the token `repo`, `workflow` and `write:packages` permissions.

Add the token as a secret to your repo under the name `PROJEN_GITHUB_TOKEN`.

### Fine-grained Personal Access Token (beta)

Follow the [GitHub docs
instructions](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-fine-grained-personal-access-token)
for creating a fine-grained personal access token (beta).

Select the repositories you want to use this token for.
You can reuse the same token for multiple repositories if you select them here.
However you should choose the minimal repository access that meets your needs.

Under Permissions, select the following Repository Permissions:

- `Contents` - Read and write
- `Metadata` - Read-only (automatically added)
- `Pull requests` - Read and write
- `Workflows` - Read and write

Add the token as a secret to your repo under the name `PROJEN_GITHUB_TOKEN`.

### GitHub App

Follow the [GitHub docs instructions](https://docs.github.com/en/developers/apps/building-github-apps/creating-a-github-app) for creating a GitHub App. Enable read & write permission for "Contents" and "Pull Request" scopes.

Add the App ID as a secret to your repo under the name `PROJEN_APP_ID` and the private key you generate as a secret to your repo under the name `PROJEN_APP_PRIVATE_KEY`.

Then, configure your projenrc file to use the GitHub app for API access:

```ts
const { github, javascript } = require('projen');

const project = new javascript.NodeProject({
  // ...other options
  githubOptions: {
    projenCredentials: github.GithubCredentials.fromApp({ ... }),
  },
});
```

If the GitHub app you are using has additional permissions assigned, you can limit the permissions used on the token within the jobs in the workflow to only allow access to write repository contents and create the pull request:

```ts
const { github, javascript } = require('projen');

const project = new javascript.NodeProject({
  // ...other options
  githubOptions: {
    projenCredentials: github.GithubCredentials.fromApp({
      // ...other options
      permissions: {
        pullRequests: github.workflows.AppPermission.WRITE,
        contents: github.workflows.AppPermission.WRITE,
      }
    }),
  },
});
```

## Workflows

See the `GitHub`, `GithubWorkflow`, and `Job` types in the [API
reference](./api/API.md) for currently available APIs.

Example code of creating a GitHub workflow:
<https://github.com/projen/projen/blob/65b4194c163f47ba4842981b0c92dbe516be787b/src/github/auto-approve.ts#L67-L105>

### Actions versions

Most workflows included with projen are constraint to a major version, in order for updates to be available immediately.
However it is a good security practice to lock versions of GitHub Actions down to explicit commit hashes.
To achieve this, you can define explicit overrides for any action used in workflows.

The replace all occurrences of an action can be overridden, irregardless of the action version:

```ts
project.github.actions.set("actions/checkout", "actions/checkout@ac59398");
```

Any use of `actions/checkout` is now changed to this:

```yaml
steps:
  - uses: actions/checkout@ac59398
```

Alternatively, any specific action version can be overridden.
This can be useful when a specific version of an action must be used due to incompatible changes.
Specific overrides take precedence over overrides without a version.

```ts
project.github.actions.set("actions/checkout@v3", "actions/checkout@ac59398");
project.github.actions.set("actions/checkout", "actions/checkout@main");
```

Different versions of `actions/checkout` are resolved to different overrides:

```yaml
steps:
  # Was: actions/checkout@v3
  - uses: actions/checkout@ac59398

  # Was: actions/checkout@v2
  - uses: actions/checkout@main
```

Workflow creators are encouraged to use commit hashes and keep them updated, or major versions.
Sensitive workflow actions should always use commit hashes.

### Stale workflow

A "stale" workflow can be added which will automatically close issues or pull
requests on your GitHub repository after time has passed without the issue
seeing any comments or updates. You can enable as shown below:

```ts
// or PythonProject, etc.
new typescript.TypeScriptProject({
  stale: true,
  staleOptions: {
    issues: {
      closeMessage: "closing pull request",
      staleLabel: "I-AM-STALE",
      daysBeforeStale: 180,
    },
  }
})
```

Check the API reference for a list of all available options.

When enabled, by default issues with no activity with will be marked as stale
after 60 days and closed within 7 days, and pull requests with no activity will
be marked as stale after 14 days and closed within 2 days.
