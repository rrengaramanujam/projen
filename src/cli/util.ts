import * as fs from "fs";
import * as path from "path";
import * as logging from "../logging";
import { exec } from "../util";

/**
 * Installs the npm module (through `npm install`) to node_modules under `projectDir`.
 * @param spec The npm package spec (e.g. `foo@^1.2` or `foo@/var/folders/8k/qcw0ls5pv_ph0000gn/T/projen-RYurCw/pkg.tgz`)
 * @returns The installed package name (e.g. `@foo/bar`)
 */
export function installPackage(baseDir: string, spec: string): string {
  const packageJsonPath = path.join(baseDir, "package.json");
  const packageJsonExisted = fs.existsSync(packageJsonPath);

  if (!packageJsonExisted) {
    // Make sure we have a package.json to read from later
    exec("npm init --yes", { cwd: baseDir });
  }

  logging.info(`installing external module ${spec}...`);
  exec(renderInstallCommand(baseDir, spec), { cwd: baseDir });

  // Get the true installed package name
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
  const packageName = Object.keys(packageJson.devDependencies).find(
    (name) => name !== "projen"
  );

  if (!packageName) {
    throw new Error(`Unable to resolve package name from spec ${spec}`);
  }

  // if package.json did not exist before calling `npm install`, we should remove it
  // so we can start off clean.
  if (!packageJsonExisted) {
    fs.rmSync(packageJsonPath, { force: true, recursive: true });
  }

  return packageName;
}

/**
 * Render a command to install an npm package.
 *
 * Engine checks are ignored at this point so that the module can be installed
 * regardless of the environment. This was needed to unblock the upgrade of the
 * minimum node version of projen, but also okay generally because engine checks
 * will be performed later and for all eternity.
 *
 * @param dir Base directory
 * @param module The module to install (e.g. foo@^1.2)
 * @returns The string that includes the install command ("npm install ...")
 */
export function renderInstallCommand(dir: string, module: string): string {
  return `npm install --save --save-dev -f --no-package-lock --prefix="${dir}" ${module}`;
}

export class CliError extends Error {
  constructor(...lines: string[]) {
    super(lines.join("\n"));
    this.name = "CliError";
  }
}
