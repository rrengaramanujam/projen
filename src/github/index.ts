export * from "./github";

export * from "./dependabot";
export * from "./workflows";
export * from "./actions-provider";
export * from "./mergify";
export * from "./pr-template";
export * from "./mergify";
export * from "./auto-merge";
export * from "./auto-approve";
export * from "./stale";
export * from "./pull-request-lint";
export * from "./github-credentials";
export * from "./workflow-actions";
export * from "./workflow-jobs";

export * from "./task-workflow";

export * from "./github-project";

export * as workflows from "./workflows-model";
