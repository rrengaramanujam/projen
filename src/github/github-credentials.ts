import { JobStep, AppPermissions } from "./workflows-model";
import { snakeCaseKeys } from "../util";

/**
 * Options for `GithubCredentials.fromPersonalAccessToken`
 */
export interface GithubCredentialsPersonalAccessTokenOptions {
  readonly secret?: string;
}

/**
 * Options for `GithubCredentials.fromApp`
 */
export interface GithubCredentialsAppOptions {
  readonly appIdSecret?: string;
  readonly privateKeySecret?: string;
  /**
   * The permissions granted to the token.
   *
   * @default - all permissions granted to the app
   */
  readonly permissions?: AppPermissions;
}

/**
 * Represents a method of providing GitHub API access for projen workflows.
 */
export class GithubCredentials {
  /**
   * Provide API access through a GitHub personal access token.
   *
   * The token must be added as a secret to the GitHub repo, and the name of the
   * secret can be specified here.
   *
   * @see https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
   * @default - a secret named "PROJEN_GITHUB_TOKEN"
   */
  public static fromPersonalAccessToken(
    options: GithubCredentialsPersonalAccessTokenOptions = {}
  ) {
    return new GithubCredentials({
      setupSteps: [],
      tokenRef: `\${{ secrets.${options.secret ?? "PROJEN_GITHUB_TOKEN"} }}`,
    });
  }

  /**
   * Provide API access through a GitHub App.
   *
   * The GitHub App must be installed on the GitHub repo, its App ID and a
   * private key must be added as secrets to the repo. The name of the secrets
   * can be specified here.
   *
   * @see https://docs.github.com/en/developers/apps/building-github-apps/creating-a-github-app
   * @see https://projen.io/github.html#github-app
   * @default - app id stored in "PROJEN_APP_ID" and private key stored in "PROJEN_APP_PRIVATE_KEY" with all permissions attached to the app
   */
  public static fromApp(options: GithubCredentialsAppOptions = {}) {
    const appIdSecret = options.appIdSecret ?? "PROJEN_APP_ID";
    const privateKeySecret =
      options.privateKeySecret ?? "PROJEN_APP_PRIVATE_KEY";

    return new GithubCredentials({
      setupSteps: [
        {
          name: "Generate token",
          id: "generate_token",
          uses: "tibdex/github-app-token@021a2405c7f990db57f5eae5397423dcc554159c",
          with: {
            app_id: `\${{ secrets.${appIdSecret} }}`,
            private_key: `\${{ secrets.${privateKeySecret} }}`,
            permissions: options.permissions
              ? JSON.stringify(snakeCaseKeys(options.permissions))
              : undefined,
          },
        },
      ],
      tokenRef: "${{ steps.generate_token.outputs.token }}",
    });
  }

  private constructor(private readonly options: GithubCredentialsOptions) {}

  /**
   * Setup steps to obtain GitHub credentials.
   */
  public get setupSteps(): JobStep[] {
    return [...this.options.setupSteps];
  }

  /**
   * The value to use in a workflow when a GitHub token is expected. This
   * typically looks like "${{ some.path.to.a.value }}".
   */
  public get tokenRef(): string {
    return this.options.tokenRef;
  }
}

interface GithubCredentialsOptions {
  readonly setupSteps: JobStep[];
  readonly tokenRef: string;
}
