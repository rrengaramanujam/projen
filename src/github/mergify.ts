import { GitHub } from "./github";
import { Component } from "../component";
import { snakeCaseKeys } from "../util";
import { YamlFile } from "../yaml";

/**
 * The Mergify conditional operators that can be used are: `or` and `and`.
 * Note: The number of nested conditions is limited to 3.
 * @see https://docs.mergify.io/conditions/#combining-conditions-with-operators
 */
export interface MergifyConditionalOperator {
  readonly or?: MergifyCondition[];
  readonly and?: MergifyCondition[];
}

export type MergifyCondition = string | MergifyConditionalOperator;

export interface MergifyRule {
  /**
   * The name of the rule. This is not used by the engine directly,
   * but is used when reporting information about a rule.
   */
  readonly name: string;
  /**
   * A list of Conditions string that must match against the
   * pull request for the rule to be applied.
   * @see https://docs.mergify.io/conditions/#conditions
   */
  readonly conditions: MergifyCondition[];
  /**
   * A dictionary made of Actions that will be executed on the
   * matching pull requests.
   * @see https://docs.mergify.io/actions/#actions
   */
  readonly actions: { [action: string]: any };
}

export interface MergifyQueue {
  /**
   * The name of the queue.
   */
  readonly name: string;

  /**
   * Merge method to use.
   *
   * Possible values are `merge`, `squash`, `rebase` or `fast-forward`.
   * `fast-forward` is not supported on queues with `speculative_checks` > 1, `batch_size` > 1, or with `allow_inplace_checks` set to false.
   *
   * @default "merge"
   */
  readonly mergeMethod?: string;

  /**
   * Method to use to update the pull request with its base branch when the speculative check is done in-place.
   *
   * Possible values:
   *  - `merge` to merge the base branch into the pull request.
   *  - `rebase` to rebase the pull request against its base branch.
   *
   * Note that the `rebase` method has some drawbacks, see Mergify docs for details.
   *
   * @see https://docs.mergify.com/actions/queue/#queue-rules
   *
   * @default - `merge` for all merge methods except `fast-forward` where `rebase` is used
   */
  readonly updateMethod?: string;

  /**
   * A list of Conditions string that must match against the
   * pull request for the pull request to be added to the queue.
   * @see https://docs.mergify.com/conditions/#conditions
   */
  readonly conditions: MergifyCondition[];
}

export interface MergifyOptions {
  readonly rules?: MergifyRule[];
  readonly queues?: MergifyQueue[];
}

export class Mergify extends Component {
  private readonly rules = new Array<MergifyRule>();
  private readonly queues = new Array<MergifyQueue>();
  // The actual YAML file will only be created if at least 1 rule is added.
  private yamlFile?: YamlFile;

  constructor(github: GitHub, options: MergifyOptions = {}) {
    super(github.project);

    for (const rule of options.rules ?? []) {
      this.addRule(rule);
    }

    for (const queue of options.queues ?? []) {
      this.addQueue(queue);
    }
  }

  private createYamlFile() {
    if (this.yamlFile == null) {
      this.yamlFile = new YamlFile(this.project, ".mergify.yml", {
        obj: {
          queue_rules: () => this.queues.map((q) => snakeCaseKeys(q, false)),
          pull_request_rules: this.rules,
        },
        // Mergify needs to read the file from the repository in order to work.
        committed: true,
      });
    }
  }

  public addRule(rule: MergifyRule) {
    this.rules.push(rule);
    this.createYamlFile();
  }

  public addQueue(queue: MergifyQueue) {
    this.queues.push(queue);
    this.createYamlFile();
  }
}
