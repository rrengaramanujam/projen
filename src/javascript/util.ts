import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}
