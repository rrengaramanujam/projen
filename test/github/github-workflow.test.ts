import * as YAML from "yaml";
import { GithubWorkflow } from "../../src/github/workflows";
import { synthSnapshot, TestProject } from "../util";

describe("github-workflow", () => {
  const workflowName = "test-workflow";

  test("Default concurrency allowed", () => {
    const project = new TestProject();

    new GithubWorkflow(project.github!, workflowName);

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).not.toContain(
      "concurrency"
    );
  });

  test("concurrency set", () => {
    const concurrencyName = "my-concurrency";
    const project = new TestProject();

    new GithubWorkflow(project.github!, workflowName, {
      concurrency: concurrencyName,
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toContain(
      `concurrency: ${concurrencyName}`
    );
  });

  test("workflow job calling a reusable workflow", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const workflow = new GithubWorkflow(project.github!, workflowName);

    workflow.addJob("call-another-workflow", {
      name: "Some reusable workflow",
      uses: "some-user/example-action.yaml@v1",
      permissions: {},
      with: {
        foo: true,
        bar: "example",
      },
      secrets: {
        credentials: "super-secret",
      },
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("workflow job calling a reusable workflow & inherited secrets", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const workflow = new GithubWorkflow(project.github!, workflowName);

    workflow.addJob("call-another-workflow", {
      name: "Some reusable workflow",
      uses: "some-user/example-action.yaml@v1",
      permissions: {},
      secrets: "inherit",
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("working-directory can be set on a Step", () => {
    const project = new TestProject();

    const ghw = new GithubWorkflow(project.github!, workflowName);
    ghw.addJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ run: "ls", workingDirectory: ".github/workflows" }],
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("workflow job calling a reusable workflow with a strategy matrix", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const workflow = new GithubWorkflow(project.github!, workflowName);

    workflow.addJob("use-a-matrix", {
      name: "Some reusable workflow",
      uses: "some-user/example-action.yaml@v1",
      permissions: {},
      secrets: "inherit",
      strategy: {
        matrix: {
          domain: {
            fruit: ["apple", "orange", "banana"],
          },
        },
      },
      with: {
        fruit: "${{ matrix.fruit }}",
      },
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("workflow job can be removed after creation", () => {
    const project = new TestProject();

    const workflow = new GithubWorkflow(project.github!, workflowName);
    workflow.addJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ uses: "actions/checkout@v3" }],
    });
    workflow.removeJob("working-dir");

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();

    const workflowYAML = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );
    expect(workflowYAML.jobs["working-dir"]).toBeUndefined();
  });

  test("workflow job can be updated after creation", () => {
    const project = new TestProject();

    const workflow = new GithubWorkflow(project.github!, workflowName);
    workflow.addJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ uses: "actions/checkout@v3" }],
    });
    workflow.updateJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ uses: "actions/checkout@v2" }],
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
    const workflowYAML = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );
    expect(workflowYAML.jobs["working-dir"].steps[0].uses).toEqual(
      "actions/checkout@v2"
    );
  });

  describe("reusable actions", () => {
    test("can use an action on a step", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [{ uses: "actions/checkout@v3" }],
      });

      const snapshot = synthSnapshot(project);

      expect(
        snapshot[`.github/workflows/${workflowName}.yml`]
      ).toMatchSnapshot();
    });

    test("can override the action used in a step", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [{ uses: "actions/checkout@v3" }],
      });

      project.github?.actions.set(
        "actions/checkout@v3",
        "replacement/checkout"
      );

      const snapshot = synthSnapshot(project);
      const wf = snapshot[`.github/workflows/${workflowName}.yml`];

      expect(wf).not.toContain("actions/checkout@v3");
      expect(wf).toContain("replacement/checkout");
    });

    test("can override action without explicitly defining version", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [{ uses: "actions/checkout@v3" }],
      });

      project.github?.actions.set("actions/checkout", "replacement/checkout");

      const snapshot = synthSnapshot(project);
      const wf = snapshot[`.github/workflows/${workflowName}.yml`];

      expect(wf).not.toContain("actions/checkout@v3");
      expect(wf).toContain("replacement/checkout");
    });

    test("can override multiple versions and the default", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [
          { uses: "actions/checkout@v2" },
          { uses: "actions/checkout@v3" },
        ],
      });

      project.github?.actions.set(
        "actions/checkout@v2",
        "actions/checkout@explicit-v2"
      );
      project.github?.actions.set(
        "actions/checkout",
        "actions/checkout@generic-override"
      );

      const snapshot = synthSnapshot(project);
      const wf = snapshot[`.github/workflows/${workflowName}.yml`];

      expect(wf).toMatchSnapshot();
      expect(wf).not.toContain("actions/checkout@v2");
      expect(wf).not.toContain("actions/checkout@v3");
      expect(wf).toContain("actions/checkout@explicit-v2");
      expect(wf).toContain("actions/checkout@generic-override");
    });
  });
});
