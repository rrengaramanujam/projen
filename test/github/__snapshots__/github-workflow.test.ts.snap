// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github-workflow reusable actions can override multiple versions and the default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - uses: actions/checkout@explicit-v2
      - uses: actions/checkout@generic-override
"
`;

exports[`github-workflow reusable actions can use an action on a step 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - uses: actions/checkout@v3
"
`;

exports[`github-workflow workflow job calling a reusable workflow & inherited secrets 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  call-another-workflow:
    name: Some reusable workflow
    permissions: {}
    uses: some-user/example-action.yaml@v1
    secrets: inherit
"
`;

exports[`github-workflow workflow job calling a reusable workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  call-another-workflow:
    name: Some reusable workflow
    permissions: {}
    uses: some-user/example-action.yaml@v1
    with:
      foo: true
      bar: example
    secrets:
      credentials: super-secret
"
`;

exports[`github-workflow workflow job calling a reusable workflow with a strategy matrix 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  use-a-matrix:
    name: Some reusable workflow
    permissions: {}
    uses: some-user/example-action.yaml@v1
    with:
      fruit: \${{ matrix.fruit }}
    secrets: inherit
    strategy:
      matrix:
        fruit:
          - apple
          - orange
          - banana
"
`;

exports[`github-workflow workflow job can be removed after creation 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs: {}
"
`;

exports[`github-workflow workflow job can be updated after creation 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - uses: actions/checkout@v2
"
`;

exports[`github-workflow working-directory can be set on a Step 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - run: ls
        working-directory: .github/workflows
"
`;
